cache/fast_lru_cache.h

```diff
@@ -117,42 +157,66 @@ class LRUHandleTable {
   explicit LRUHandleTable(int hash_bits);
   ~LRUHandleTable();
 
+  // Returns a pointer to the handle matching the key/hash, or nullptr
+  // if not present.
   LRUHandle* Lookup(const Slice& key, uint32_t hash);
-  LRUHandle* Insert(LRUHandle* h);
+  // Returns a pointer to the inserted handle, or nullptr if no slot
+  // available was found. The output argument old is set to a
+  // pointer to an existing handle that matches the key/hash, or nullptr
+  // if none is found.
+  LRUHandle* Insert(LRUHandle* h, LRUHandle** old);
+  // Returns a pointer to the handle matching the key/hash, or nullptr
+  // if not present. Assumes the element is in the table.
   LRUHandle* Remove(const Slice& key, uint32_t hash);
 
+  void Exclude(LRUHandle* e);
+
+  void Assign(int slot, LRUHandle* src);
+
   template <typename T>
   void ApplyToEntriesRange(T func, uint32_t index_begin, uint32_t index_end) {
     for (uint32_t i = index_begin; i < index_end; i++) {
-      LRUHandle* h = list_[i];
-      while (h != nullptr) {
-        auto n = h->next_hash;
-        assert(h->InCache());
+      LRUHandle* h = &array_[i];
+      if (h->IsVisible()) {
         func(h);
-        h = n;
       }
     }
   }
 
   int GetLengthBits() const { return length_bits_; }
 
-  // Return the address of the head of the chain in the bucket given
-  // by the hash.
-  inline LRUHandle** Head(uint32_t hash);
+  uint32_t GetOccupancy() const { return occupancy_; }
 
  private:
-  // Return a pointer to slot that points to a cache entry that
-  // matches key/hash.  If there is no such cache entry, return a
-  // pointer to the trailing slot in the corresponding linked list.
-  LRUHandle** FindPointer(const Slice& key, uint32_t hash);
+  int FindElement(const Slice& key, uint32_t hash, int& probe,
+                  int displacement);
+
+  int FindPresentElement(const Slice& key, uint32_t hash, int& probe,
+                         int displacement);
+
+  int FindAvailableSlot(const Slice& key, int& probe, int displacement);
+
+  int FindPresentElementOrAvailableSlot(const Slice& key, uint32_t hash,
+                                        int& probe, int displacement);
+
+  // Returns the index of the first slot probed that has a handle e
+  // such that cond(e) is true.  If the probe sequence ends without
+  // a match, returns a pointer to the first slot available in the
+  // sequence.  Otherwise, if no match is found and no slot is
+  // available, returns -1.  For every handle e probed, except the
+  // matching slot, we update e->displacements += displacement.
+  // The variable probe is modified such that consecutive calls to
+  // FindSlot continue probing where the previous left.
+  int FindSlot(const Slice& key, std::function<bool(LRUHandle*)> cond,
```

Comment:
How about use Status as the return value for these new functions?

Reply:
The design rule I followed here is that when the operation is internal to the shard, I don't use Status. For example, the other LRUHandleTable's public functions report status with nullptr (e.g., Insert returns nullptr if no slot is available).

Reply:
I agree with avoiding Status in file-local, performance critical functions